### PR TITLE
Create minimum settings file if it doesn't exist

### PR DESCRIPTION
**Pull request by wigging**
This creates a settings.yaml file if it does not exist. This is the first step to improve the settings/configuration of the flowcept package. Resolves issue #119 .

### CODE REVIEW

File: src/flowcept/configs.py [hunk 1]
```diff
@@ -12,21 +13,24 @@
...
-        f"environment variable FLOWCEPT_SETTINGS_PATH with its path or "
-        f"install Flowcept's package to create the directory "
-        f"~/.flowcept with the file in it.\n"
-        "A sample settings file is found in the 'resources' directory "
-        "under the project's root path."
-    )
+    os.makedirs(SETTINGS_DIR)
+
+    with open(SETTINGS_PATH, "w") as f:
+        base_settings = """\
```
Comment (renan-souza) [inline src/flowcept/configs.py:23]: Please do not follow this approach. As we discussed on Teams, Flowcept should work even without any settings file. We do not need to create one to make flowcept work.

Understand that this is what the software currently expects, but this is not the right behavior.

Comment (renan-souza) [reply src/flowcept/configs.py:23]: I created an issue https://github.com/ORNL/flowcept/issues/161 to organize the requirements for this, based on what we discussed on Teams.

Comment (wigging) [reply src/flowcept/configs.py:23]: This resolves issue #119. I will get to the better solution next.

Comment (wigging) [reply src/flowcept/configs.py:23]: Flowcept is broken because it does not work without a settings file. If you pip install flowcept from PyPI (not from the GitHub repository) then it will not work on the users computer. This pull request is a simple solution to fix that problem while using the current config module.

Comment (wigging) [reply src/flowcept/configs.py:23]: Yes, a better solution exists. But this is the quickest way to fix the currently broken package on PyPI.

Comment (renan-souza) [reply src/flowcept/configs.py:23]: > Flowcept is broken because it does not work without a settings file.
Flowcept was initially designed to always work with a settings file. Based on this, Flowcept is not broken. It is working according to this original design. We are currently in the process of refining this design, to allow it to work without a settings file.

Addressing issue https://github.com/ORNL/flowcept/issues/161 entirely will resolve https://github.com/ORNL/flowcept/issues/119.

Let's go after https://github.com/ORNL/flowcept/issues/161, beginning with enable Flowcept to run without a settings file.

Comment (renan-souza) [reply src/flowcept/configs.py:23]: > > Flowcept was initially designed to always work with a settings file. Based on this, Flowcept is not broken.
> 
> If it's not broken then why does it not work when you pip install it from PyPI?

If you carefully read the configs.py file, you will understand that the way it is currently implemented, it works (including if you pip install it from pypi) but you will be required to define the env var `FLOWCEPT_SETTINGS_PATH` to a valid yaml file.

By valid I mean that it **now** must contain exactly what the configs.py expects it to be. 

What I am proposing is to make the utilization of Flowcept more flexible and straightforward, without requiring extra steps to get it running with the most simple example.

Going back to that set of four steps that I would like to enable (see https://github.com/ORNL/flowcept/issues/161), currently, one extra step is needed, i.e., to define the env var pointing to a valid yaml file. We want to remove this extra step in this issue #161, making it easier for the user.

In summary, it works perfectly if you pip install it from pypi, but it requires an unnecessary extra step to deal with the settings file.

Comment (wigging) [reply src/flowcept/configs.py:23]: I feel like I'm shouting into a void here.

> it will work if you pip install it from pypi but you will be required to define the env var FLOWCEPT_SETTINGS_PATH to point to valid yaml file

Again, it will not work if you pip install it from PyPI. You cannot assume that `FLOWCEPT_SETTINGS_PATH` has already been defined as an environment variable. You cannot assume that a valid settings file already exists at that path. If these things are required then they must be implemented by the package itself. Not as some extra steps by the user.

> I always keep a valid default yaml file under the directory resources.

This `resources` directory is in the GitHub repository, not the package itself. You cannot rely on the user to go to some repository for extra files that are needed to use the package. Any reputable package will automatically install the necessary files to use the package. If any file is needed for a package to run, then that file or a way to generate that file must be implemented upon installation of the package.

> In summary, it works perfectly if you pip install it from pypi, but it requires an unnecessary extra step to deal with the settings file

The fact that it requires extra steps to deal with the settings file is exactly why it is broken. It cannot be used without a settings file. You must do all the things mentioned above to point to a valid settings file and/or create that settings file. The current version of the flowcept package does none of these things. Again, there are no reputable Python packages that require you to do "extra steps" to get them to work. Sphinx and Django use settings (config) files but they don't require the user to go somewhere to get that settings/config file. They automatically create that file for the user.

Yes, submitting changes that resolve issue #161 will fix these problems. But these changes may take time to properly execute. So until that happens, this pull request provides an immediate solution which will automatically generate a minimum settings file for the user. Consequently, users can install from PyPI without having to do these extra steps for the settings file.

Comment (renan-souza) [reply src/flowcept/configs.py:23]: Ok, let me do a proper quick fix which will solve the #161-3 and #119 .

Do you mind if I do it on this branch (and PR)?

Comment (wigging) [reply src/flowcept/configs.py]: For this pull request, I didn't get rid of the settings file because:
- There seems to be other areas of the package (other than configs.py) that rely on the file. It's hard to tell where this occurs because the other files in the package import various things from configs.py. It would be easier to troubleshoot if the whole config module was imported instead of individual variables.
- The logging functionality relies on the presence of the directory that the settings file resides in. This is related to the SETTINGS_PATH and SETTINGS_DIR on config.py. So not having that directory will also break flowcept.

Anyway, the solution I pushed up for this pull request is just the simplest thing I could come up with. I understand (and agree) that not requiring a settings file is the desired solution. But for now I just wanted to push up a quick fix so that at least the PyPI install would work without requiring the user to manually create a settings file and the associated directory for that file.

Comment (wigging) [reply src/flowcept/configs.py]: Also, using `os` is not a good approach to work with files and directories. It was fine in the Python 2 and early Python 3 days. These days the `pathlib` module is preferred over `os` because it has better support for cross-platform file systems and is easier to use.

Comment (renan-souza) [reply src/flowcept/configs.py]: Your quick fix did not work. The CI tests were copying the sample_settings.yaml to the folder ~/<user>/.flowcept. That is why the tests passed. If you'd removed this step from the CI tests, they would've failed. 

I did a quick fix, which I will explain in detail in issue #161

**Comment by wigging** [comment]
Please stop merging pull requests that I create!

File: .github/workflows/run-tests.yml [hunk 2]
```diff
@@ -20,23 +20,31 @@ jobs:
...
-      - name: Copy settings file
-        run: |
-          mkdir ~/.flowcept
-          cp resources/sample_settings.yaml ~/.flowcept
-          mv ~/.flowcept/sample_settings.yaml ~/.flowcept/settings.yaml
+#      - name: Copy settings file
+#        run: |
+#          mkdir ~/.flowcept
+#          cp resources/sample_settings.yaml ~/.flowcept
+#          mv ~/.flowcept/sample_settings.yaml ~/.flowcept/settings.yaml
```
Comment (wigging) [inline .github/workflows/run-tests.yml:27]: Just remove this from the file. Leaving it as commits just clutters up the file.

File: pyproject.toml [hunk 3]
```diff
@@ -102,3 +102,9 @@ ignore = ["D200", "D212"]
 
 [tool.ruff.lint.pydocstyle]
 convention = "numpy"
+
+[tool.hatch.build.targets.wheel]
+packages = ["src/flowcept"]
```
Comment (wigging) [inline pyproject.toml:107]: Why are you telling the build tool (hatch) what the package directory is? The src layout is automatically recognized by hatch and shouldn't need to be explicitly defined. Did you try installing without these lines?

File: pyproject.toml [hunk 4]
```diff
@@ -102,3 +102,9 @@ ignore = ["D200", "D212"]
...
+
+[tool.hatch.build.targets.wheel.force-include]
+"resources/sample_settings.yaml" = "resources/sample_settings.yaml"
```
Comment (wigging) [inline pyproject.toml:110]: If you are going to make the settings file required by the package then just put it in the source directory of the package. Also, I don't see any changes made to this sample_settings.yaml file. I think it had adapters defined for zambeze, mlflow, tensorboard, and dask which means all those dependencies would also be needed for the installation. I thought the goal was to provide a minimum settings file with no adapters?
